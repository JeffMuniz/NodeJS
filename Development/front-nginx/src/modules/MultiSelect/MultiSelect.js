import React from "react";
import PropTypes, { string } from "prop-types";
import { isEmpty, isInteger, isBoolean, some } from "lodash";
import "react-select/dist/react-select.css";

import { Select } from "./MultiSelect.styles";

const getValue = value => {
  const isEmptyObj = some(value, field => {
    if (isInteger(field) || isBoolean(field)) return false;
    return isEmpty(field);
  });
  return isEmptyObj ? "" : value;
};

const MultiSelect = props => {
  const {
    items,
    multiSelect,
    selectedItem,
    getMoreData,
    isCapitalized,
    maxMenuHeight,
  } = props;
  if (multiSelect && !items.some(item => item.id === 0)) {
    items.push({ id: 0, name: "Todos" });
  }

  return (
    <Select
      options={items.sort((a, b) => a.id > b.id)}
      multi={multiSelect}
      value={getValue(selectedItem)}
      valueKey="id"
      labelKey="name"
      onMenuScrollToBottom={getMoreData}
      clearable={false}
      isCapitalized={isCapitalized}
      maxMenuHeight={maxMenuHeight}
      {...props}
    />
  );
};

export default MultiSelect;

MultiSelect.propTypes = {
  id: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  getMoreData: PropTypes.func,
  multiSelect: PropTypes.bool,
  hasError: PropTypes.bool,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      id: PropTypes.number,
    }),
  ),
  placeholder: PropTypes.string,
  selectedItem: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.shape({
      name: PropTypes.string,
      id: PropTypes.number,
    }),
  ]),
  isCapitalized: PropTypes.bool,
  maxMenuHeight: string,
};

MultiSelect.defaultProps = {
  multiSelect: false,
  hasError: false,
  onChange: () => null,
  onBlur: () => null,
  getMoreData: () => null,
  placeholder: "Selecione...",
  items: [],
  selectedItem: null,
  isCapitalized: false,
  maxMenuHeight: "",
};
