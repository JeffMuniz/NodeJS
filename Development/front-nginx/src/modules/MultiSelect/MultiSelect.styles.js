import ReactSelect from "react-select";
import styled from "styled-components";
import { red, darkGrey, white } from "src/styles/colors";

export const Select = styled(ReactSelect)`
  & .Select-control {
    background-color: ${white};
    border-radius: 8px;
    border: ${props =>
      props.hasError ? `1px solid ${red}` : `1px solid ${darkGrey}`};
    border-spacing: 0;
    border-collapse: separate;
    box-sizing: border-box;
    -moz-box-sizing: border-box;
    -webkit-box-sizing: border-box;
    color: #3c3b3b;
    display: table;
    font-weight: 100;
    font-size: 14px;
    height: 40px;
    outline: none;
    overflow: auto;
    z-index: 9999;
  }
  &.is-disabled.Select > .Select-control {
    background-color: rgba(248, 248, 248, 0.5);
  }
  & .Select-placeholder {
    line-height: 40px;
  }
  &.Select--single > .Select-control .Select-value {
    line-height: 40px;
  }

  & .Select-control > .Select-multi-value-wrapper {
    max-height: 50px;
  }

  & .Select-menu {
    max-height: ${props =>
      props.maxMenuHeight ? props.maxMenuHeight : "198px"};
  }
`;
