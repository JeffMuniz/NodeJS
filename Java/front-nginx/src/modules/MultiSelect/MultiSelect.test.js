import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import TestRenderer from "react-test-renderer";
import MultiSelect from "./MultiSelect";

let defaultProps = {};
let testRenderer = null;
let instance = null;

beforeEach(() => {
  defaultProps = {
    id: "select",
    items: [
      { name: "test", id: 1 },
      { name: "test2", id: 2 },
    ],
    onChange: jest.fn(),
    onBlur: jest.fn(),
    getMoreData: jest.fn(),
  };

  testRenderer = TestRenderer.create(<MultiSelect {...defaultProps} />);
  instance = testRenderer.root;
});

describe("MultiSelect", () => {
  it("Should render MultiSelect", () => {
    // given
    const renderer = new ShallowRenderer();
    const props = {
      selectedItem: { name: "test", id: 1 },
    };

    // when
    renderer.render(<MultiSelect {...defaultProps} {...props} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });

  it("Should render MultiSelect without selectedItem", () => {
    // given
    const renderer = new ShallowRenderer();
    const props = {
      selectedItem: { name: "" },
    };
    // when
    renderer.render(<MultiSelect {...defaultProps} {...props} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });

  it("Should return null when call onChange of default props", () => {
    // given
    const props = { id: "test", name: "name" };
    testRenderer.update(<MultiSelect {...props} />);

    // when
    const result = instance.props.onChange();

    // then
    expect(result).tomacull();
  });

  it("Should return null when call onBlur of default props", () => {
    // given
    const props = { id: "test", name: "name" };
    testRenderer.update(<MultiSelect {...props} />);

    // when
    const result = instance.props.onBlur();

    // then
    expect(result).tomacull();
  });

  it("Should return null when call getMoreData of default props", () => {
    // given
    const props = { id: "test", name: "name" };
    testRenderer.update(<MultiSelect {...props} />);

    // when
    const result = instance.props.getMoreData();

    // then
    expect(result).tomacull();
  });
});
