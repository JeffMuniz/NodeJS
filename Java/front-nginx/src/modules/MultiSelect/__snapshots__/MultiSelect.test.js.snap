// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect Should render MultiSelect 1`] = `
<Styled(Select)
  clearable={false}
  getMoreData={[MockFunction]}
  hasError={false}
  id="select"
  isCapitalized={false}
  items={
    Array [
      Object {
        "id": 1,
        "name": "test",
      },
      Object {
        "id": 2,
        "name": "test2",
      },
    ]
  }
  labelKey="name"
  maxMenuHeight=""
  multi={false}
  multiSelect={false}
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  onMenuScrollToBottom={[MockFunction]}
  options={
    Array [
      Object {
        "id": 1,
        "name": "test",
      },
      Object {
        "id": 2,
        "name": "test2",
      },
    ]
  }
  placeholder="Selecione..."
  selectedItem={
    Object {
      "id": 1,
      "name": "test",
    }
  }
  value={
    Object {
      "id": 1,
      "name": "test",
    }
  }
  valueKey="id"
/>
`;

exports[`MultiSelect Should render MultiSelect without selectedItem 1`] = `
<Styled(Select)
  clearable={false}
  getMoreData={[MockFunction]}
  hasError={false}
  id="select"
  isCapitalized={false}
  items={
    Array [
      Object {
        "id": 1,
        "name": "test",
      },
      Object {
        "id": 2,
        "name": "test2",
      },
    ]
  }
  labelKey="name"
  maxMenuHeight=""
  multi={false}
  multiSelect={false}
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
  onMenuScrollToBottom={[MockFunction]}
  options={
    Array [
      Object {
        "id": 1,
        "name": "test",
      },
      Object {
        "id": 2,
        "name": "test2",
      },
    ]
  }
  placeholder="Selecione..."
  selectedItem={
    Object {
      "name": "",
    }
  }
  value=""
  valueKey="id"
/>
`;
